File: Problems/No.2309_Greatest_English_Letter_in_Upper_and_Lower_Case/greatest_english_letter_in_upper_and_lower_case.py
class Solution(object):
    def greatestLetter(self, s):
        """
        :type s: str
        :rtype: str
        """

        # parameter
        letter_set = set(s)

        # start computing
        upper, lower = ord('Z'), ord('z')
        for i in range(26):
            if ((chr(upper-i) in letter_set) and (chr(lower-i) in letter_set)):
                return chr(upper-i)
        
        return ""

File: Problems/No.2309_Greatest_English_Letter_in_Upper_and_Lower_Case/greatest_english_letter_in_upper_and_lower_case.cpp
class Solution {
public:
    string greatestLetter(string s) {
        
        // parameters
        char greatest_letter = '@';
        std::unordered_set<char> letter_set;
        
        // start computing
        for (auto &c: s) {

            if (!std::isalpha(c)) continue;
            if (std::islower(c)) {

                char upper_c = std::toupper(c);
                if (letter_set.find(upper_c) != letter_set.end())
                    greatest_letter = (upper_c > greatest_letter) ? upper_c : greatest_letter;

            }
            else if (letter_set.find(std::tolower(c)) != letter_set.end())
                    greatest_letter = (c > greatest_letter) ? c : greatest_letter;

            letter_set.insert(c);

        }

        return (greatest_letter != '@') ? std::string (1, greatest_letter) : "";

    }
};
